bindings/rust/s2n-tls/src/fingerprint.rs

```diff
@@ -0,0 +1,675 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Methods to fingerprint ClientHellos.
+//!
+//! See [the C API documentation](https://github.com/aws/s2n-tls/blob/main/api/unstable/fingerprint.h).
+
+use crate::{
+    client_hello::ClientHello,
+    error::{Error, Fallible},
+    ffi::*,
+};
+
+use core::ptr::NonNull;
+
+#[non_exhaustive]
+#[derive(Copy, Clone)]
+pub enum FingerprintType {
+    JA3,
+}
+
+impl From<FingerprintType> for s2n_tls_sys::s2n_fingerprint_type::Type {
+    fn from(value: FingerprintType) -> Self {
+        match value {
+            FingerprintType::JA3 => s2n_tls_sys::s2n_fingerprint_type::FINGERPRINT_JA3,
+        }
+    }
+}
+
+/// A fingerprint operation.
+pub struct Fingerprint<'a>(&'a mut Builder);
```

Comment:
What do you think about adding an example for this in `rust-examples`? I think the current solution is quite neat (having the Builder own the memory), but am a little bit worried about ensnaring ourselves in lifetimes. It works out in my head, but I'd feel a bit more comfortable if we had an example. (e.g. log the fingerprints of 100 incoming client connections).

Reply:
I'm not opposed to it, but I don't think it resolves any lifetime concerns and I don't think it needs to be part of this PR. I have several tests and doc examples showing the lifetimes work out. I added another one specifically to test connections, like an example would.